src/bootstrap/jobs/mod.rs

```diff
@@ -11,3 +11,86 @@ pub mod http_tracker;
 pub mod torrent_cleanup;
 pub mod tracker_apis;
 pub mod udp_tracker;
+
+/// This is the message that the "launcher" spawned task sends to the main
+/// application process to notify the service was successfully started.
+///
+#[derive(Debug)]
+pub struct Started {
+    pub address: std::net::SocketAddr,
+}
+
+pub async fn make_rust_tls(enabled: bool, cert: &Option<String>, key: &Option<String>) -> Option<Result<RustlsConfig, Error>> {
+    if enabled {
+        if let (Some(cert), Some(key)) = (cert, key) {
+            info!("Using https: cert path: {cert}.");
+            info!("Using https: key path: {cert}.");
+
+            Some(
+                RustlsConfig::from_pem_file(cert, key)
+                    .await
+                    .map_err(|err| Error::BadTlsConfig {
+                        source: (Arc::new(err) as DynError).into(),
+                    }),
+            )
+        } else {
+            Some(Err(Error::MissingTlsConfig {
+                location: Location::caller(),
+            }))
+        }
+    } else {
+        info!("tls not enabled");
+        None
+    }
+}
```

Comment:
Hi @da2ce7 I think there are too many nested conditions.In general I prefer early return on errors:

```rust
pub async fn make_rust_tls(enabled: bool, cert: &Option<String>, key: &Option<String>) -> Option<Result<RustlsConfig, Error>> {
    if !enabled {
        info!("tls not enabled");
        return None;
    }

    match (cert, key) {
        (Some(cert), Some(key)) => {
            info!("Using HTTPS: cert path: {cert}, key path: {key}.");

            Some(
                RustlsConfig::from_pem_file(cert, key)
                    .await
                    .map_err(|err| Error::BadTlsConfig {
                        source: (Arc::new(err) as DynError).into(),
                    }),
            )
        }
        _ => Some(Err(Error::MissingTlsConfig {
            location: Location::caller(),
        })),
    }
}
```

I would like to also return if the `cert` or the `key` are missing but then you would need an `unwrap`:

```rust
pub async fn make_rust_tls(enabled: bool, cert: &Option<String>, key: &Option<String>) -> Option<Result<RustlsConfig, Error>> {
    if !enabled {
        info!("TLS not enabled");
        return None;
    }

    if cert.is_none() || key.is_none() {
        return Some(Err(Error::MissingTlsConfig {
            location: Location::caller(),
        }));
    }

    let cert = cert.as_ref().unwrap(); // Safe to unwrap as we checked above
    let key = key.as_ref().unwrap(); // Safe to unwrap as we checked above

    info!("Using HTTPS: cert path: {cert}, key path: {key}.");

    RustlsConfig::from_pem_file(cert, key)
        .await
        .map_err(|err| Error::BadTlsConfig {
            source: Arc::new(err).into(),
        })
        .into()
}
```

Besides, this makes me wonder if we are doing this check in the right place. Maybe we should not even accept the invalid configuration. In the configuration, we have:

```rust
pub struct HttpTracker {
    /// Weather the HTTP tracker is enabled or not.
    pub enabled: bool,
    /// The address the tracker will bind to.
    /// The format is `ip:port`, for example `0.0.0.0:6969`. If you want to
    /// listen to all interfaces, use `0.0.0.0`. If you want the operating
    /// system to choose a random port, use port `0`.
    pub bind_address: String,
    /// Weather the HTTP tracker will use SSL or not.
    pub ssl_enabled: bool,
    /// Path to the SSL certificate file. Only used if `ssl_enabled` is `true`.
    #[serde_as(as = "NoneAsEmptyString")]
    pub ssl_cert_path: Option<String>,
    /// Path to the SSL key file. Only used if `ssl_enabled` is `true`.
    #[serde_as(as = "NoneAsEmptyString")]
```

And there is a data structure that is repeated in the API, and HTTP trackers:

```rust
    pub ssl_enabled: bool,
    /// Path to the SSL certificate file. Only used if `ssl_enabled` is `true`.
    #[serde_as(as = "NoneAsEmptyString")]
    pub ssl_cert_path: Option<String>,
    /// Path to the SSL key file. Only used if `ssl_enabled` is `true`.
    #[serde_as(as = "NoneAsEmptyString")]
```

Maybe we should extract a new struct that enforces the configuration rule:

Rule: if SSL is enabled SSL cert and SSL Key are mandatory:

```rust
pub struct TLSConfig {
    pub ssl_enabled: bool,
    /// Path to the SSL certificate file. Only used if `ssl_enabled` is `true`.
    pub ssl_cert_path: Option<String>,
    /// Path to the SSL key file. Only used if `ssl_enabled` is `true`.
    pub ssl_key_path: Option<String>,
}

impl TLSConfig {
    pub fn new(ssl_enabled: bool, ssl_cert_path: Option<String>, ssl_key_path: Option<String>) -> Result<Self, &'static str> {
        if ssl_enabled && (ssl_cert_path.is_none() || ssl_key_path.is_none()) {
            Err("Both cert and key must be provided when SSL is enabled")
        } else {
            Ok(TLSConfig {
                ssl_enabled,
                ssl_cert_path,
                ssl_key_path,
            })
        }
    }
}
```

That would make parsing configuration harder but we can keep the flat structure for the toml file and add a new method to `HttpTracker` to return the `TLSConfig`.

Alternatively, we could add a validate method to all configuration structs. I guess the configuration structs are only DTOs which contain the same information as the toml file, but that does not mean the configuration is valid. Maybe we should have two types of structs for configuration:

- The DTO: the way we transfer configuration to JSON, toml, etc.
- The actual domain config structs that are validated. You cannot instantiate invalid configuration (which is possible right now).

I think we should do something similar to what Nginx does:

We could statically validate a configuration file without running it. With the current solution, we need to run the tracker with a given configuration to know if the configuration is valid.

The only validation we are doing now is the parsing process. WE try to deserialize the toml file into Rust structs with certain types. But for example, socket addresses are of type `String`:

```rust
pub bind_address: String,
```

That confirms that a second level of validation and a higher-level of domain structs 
for configuration could be a nice thing.

- Higer-level: domain configuration structs. Reach types and valid config with invariants.
- Lower-level: DTOs for configuration to be able to represent the configuration in different formats: toml, JSON, yaml.